<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Quantitative_ChIPseq_Workshop • QuantitativeChIPseqWorkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Quantitative_ChIPseq_Workshop">
<meta property="og:description" content="This package encapsulates a workshop demonstrating the quantitative analysis (especially differential binding) of ChIP-seq, ATAC-seq and related assays using R packages (particularly DiffBind).">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">QuantitativeChIPseqWorkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Quantitative-ChIPseq-Workshop.html">Workshop: Quantitative analysis of ChiP-seq, ATAC-seq, and related DNA enrichment assays</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bioinformatics-core-shared-training/Quantitative-ChIPseq-Workshop">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level1">
<div class="page-header"><h1 id="quantitiative-analysis-of-chip-seq-atac-seq-and-related-dna-enrichment-assays">Quantitiative analysis of ChiP-seq, ATAC-seq, and related DNA enrichment assays.<a class="anchor" aria-label="anchor" href="#quantitiative-analysis-of-chip-seq-atac-seq-and-related-dna-enrichment-assays"></a>
</h1></div>
</div>
<div class="section level1">
<h1 id="instructor">Instructor<a class="anchor" aria-label="anchor" href="#instructor"></a>
</h1>
<p>Rory Stark, Principal Scientist Computational Biology, University of Cambridge, Cancer Research UK Cambridge Institute</p>
<p><a href="mailto:rory.stark@cruk.cam.ac.uk" class="email">rory.stark@cruk.cam.ac.uk</a></p>
</div>
<div class="section level1">
<h1 id="workshop-description">Workshop Description<a class="anchor" aria-label="anchor" href="#workshop-description"></a>
</h1>
<p>This workshop will demonstrate the steps involved in performing a quantitative analysis of ChIP-seq data in <em>Bioconductor</em> (specifically differential binding analysis), with some discussion of related assays such as ATAC-seq. Particular attention will be paid to processing of aligned reads, including blacklisting, greylisting, filtering for quality and duplication, and the particular challenges presented when normalizing these data. While the workshop follows the <em>DiffBind</em> package vignette, the use of a number of other <em>Bioconductor</em> packages is discussed, including <em>csaw</em>, <em>ChIPQC</em>, <em>edgeR</em>, <em>DESeq2</em>, and <em>GreyListChIP</em>.</p>
<div class="section level2">
<h2 id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h2>
<p>Attendees should be familiar with the following:</p>
<ul>
<li>Basic knowledge of R syntax</li>
<li>Familiarity with core <em>Bioconductor</em> classes, such as <em>GRanges</em>.</li>
<li>Familiarity with ChIP-seq or a related assay, including aligning sequencing reads.</li>
<li>Familiarity with generalized linear models (GLMs), especially as realized in the <em>edgeR</em> and/or <em>DESeq2</em> packages.</li>
</ul>
<p>Helpful background reading:</p>
<ul>
<li>Read through the <em>DiffBind</em> package vignette</li>
<li>Read through the <em>csaw</em> package vignette</li>
</ul>
<p>There is a book chapter which offers a detailed outline of the design and analysis of ChIP-seq experiments:</p>
<ul>
<li>Stark, Rory, and James Hadfield. “Characterization of DNA-protein interactions: design and analysis of ChIP-seq experiments.” <em>Field Guidelines for Genetic Experimental Designs in High-Throughput Sequencing.</em> Springer, Cham, 2016. 223-260.</li>
</ul>
<p>Email the instructor if you would like a copy.</p>
</div>
<div class="section level2">
<h2 id="workshop-participation">Workshop participation<a class="anchor" aria-label="anchor" href="#workshop-participation"></a>
</h2>
<p>Students will participate by following along an <a href="https://rmarkdown.rstudio.com/" class="external-link"><em>R markdown</em></a> document, and asking questions throughout the workshop.</p>
<p>The conference organizers have arranged for attendees to have access to a RStudio session running inside a Docker instance that contains everything you need to follow the workshop.</p>
<p><strong>NOTE</strong>: <em>The workshop Docker instance does not have much compute power.</em> <em>As a result, it is not possible to run all the code in the time allowed.</em> <em>Two operations are affected: building greylists, and counting reads.</em> <em>In these cases, the workshop includes pre-computed data objects that</em> <em>can be loaded to avoid the time-consuming code.</em> <em>This is noted inline.</em></p>
</div>
<div class="section level2">
<h2 id="workshop-participation--local-docker-image">Workshop participation – local Docker image<a class="anchor" aria-label="anchor" href="#workshop-participation--local-docker-image"></a>
</h2>
<p>If you do <strong>not</strong> have access to the RStudio instance provided by the conference organizers, or you would like to run the Docker image locally and you have a Docker client installed on your machine, you run this workshop locally via:</p>
<pre><code>docker pull crukcibioinformatics/quantitative_chip_workshop:latest
docker images
docker run -e PASSWORD=DiffBind -p 8787:8787 &lt;imageid&gt; </code></pre>
<p>Then open a browser and go to the URL <strong>localhost:8787</strong>. Log into RStudio with <strong>username:rstudio</strong> and <strong>password:DiffBind</strong>.</p>
</div>
<div class="section level2">
<h2 id="workshop-participation--local-r-installation">Workshop participation – local R installation<a class="anchor" aria-label="anchor" href="#workshop-participation--local-r-installation"></a>
</h2>
<p>The workshop uses a Docker container with Bioconductor Release version <code>3.12</code>. If you would like to install Bioconductor on your computer, install R and the required Bioconductor packages, including this workshop package.</p>
</div>
<div class="section level2">
<h2 id="how-to-follow-the-workshop">How to follow the workshop<a class="anchor" aria-label="anchor" href="#how-to-follow-the-workshop"></a>
</h2>
<p>First load the workshop package:</p>
<pre><code><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bioinformatics-core-shared-training/Quantitative-ChIPseq-Workshop">QuantitativeChIPseqWorkshop</a></span><span class="op">)</span></code></pre>
<p>There are a number of ways to follow the workshop:</p>
<div class="section level3">
<h3 id="follow-the-workshop-in-the-pre-built-html-in-rstudio">Follow the workshop in the pre-built HTML in RStudio<a class="anchor" aria-label="anchor" href="#follow-the-workshop-in-the-pre-built-html-in-rstudio"></a>
</h3>
<p>You can open the fully built workshop (with all code and results) in the “Help” pane in RStudio as follows:</p>
<pre><code><span class="fu"><a href="https://rdrr.io/r/utils/vignette.html" class="external-link">vignette</a></span><span class="op">(</span><span class="st">"Quantitative-ChIPseq-Workshop"</span>,<span class="st">"QuantitativeChIPseqWorkshop"</span><span class="op">)</span></code></pre>
</div>
<div class="section level3">
<h3 id="follow-the-workshop-in-the-pre-built-html-in-a-separate-browser-window-or-tab">Follow the workshop in the pre-built HTML in a separate Browser window or tab<a class="anchor" aria-label="anchor" href="#follow-the-workshop-in-the-pre-built-html-in-a-separate-browser-window-or-tab"></a>
</h3>
<p>If you are using the Docker image, You can open the fully built workshop (with all code and results) in a browser tab as follows:</p>
<pre><code><span class="fu">rstudioapi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rstudioapi/man/viewer.html" class="external-link">viewer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"doc"</span>,package<span class="op">=</span><span class="st">"QuantitativeChIPseqWorkshop"</span><span class="op">)</span>,
                         <span class="st">"Quantitative-ChIPseq-Workshop.html"</span>, sep<span class="op">=</span><span class="st">"/"</span><span class="op">)</span><span class="op">)</span></code></pre>
<p>It may be helpful to move the tab into its own window so you can see the vignette and the RStudio window at the same time.</p>
</div>
<div class="section level3">
<h3 id="follow-the-workshop-directly-in-the-workshop-mark-down-file">Follow the workshop directly in the workshop mark-down file<a class="anchor" aria-label="anchor" href="#follow-the-workshop-directly-in-the-workshop-mark-down-file"></a>
</h3>
<p>You can open the workshop “source” .Rmd file in the RStudio “Source” pane, and follow along, executing each code chunk in sequence, to build it as you go.</p>
<pre><code><span class="fu"><a href="https://rdrr.io/r/utils/file.edit.html" class="external-link">file.edit</a></span><span class="op">(</span><span class="st">"vignettes/Quantitative-ChIPseq-Workshop.Rmd"</span><span class="op">)</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="r--bioconductor-packages-used">
<em>R</em> / <em>Bioconductor</em> packages used<a class="anchor" aria-label="anchor" href="#r--bioconductor-packages-used"></a>
</h2>
<ul>
<li><em>DiffBind</em></li>
<li><em>csaw</em></li>
<li><em>ChIPQC</em></li>
<li><em>DESeq2</em></li>
<li><em>edgeR</em></li>
<li><em>GreyListChIP</em></li>
</ul>
</div>
<div class="section level2">
<h2 id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h2>
<p>The workshop duration is 90 min. Approximate timing of activities:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Introduction</td>
<td>5m</td>
</tr>
<tr class="even">
<td>QC of DNA enrichment assays</td>
<td>5m</td>
</tr>
<tr class="odd">
<td>Peaks vs. Windows</td>
<td>5m</td>
</tr>
<tr class="even">
<td>
<em>DiffBind</em> Sample sheet</td>
<td>5m</td>
</tr>
<tr class="odd">
<td>Blacklists and Greylists</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Counting aligned reads</td>
<td>10m</td>
</tr>
<tr class="odd">
<td>Examining and normalizing the binding matrix</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Modeling and Testing</td>
<td>5m</td>
</tr>
<tr class="odd">
<td>Examining analysis results</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Multi-factor Design</td>
<td>5m</td>
</tr>
<tr class="odd">
<td>Normalization discussion</td>
<td>15m</td>
</tr>
<tr class="even">
<td>Conclusions</td>
<td>5m</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level1">
<h1 id="workshop-goals-and-objectives">Workshop goals and objectives<a class="anchor" aria-label="anchor" href="#workshop-goals-and-objectives"></a>
</h1>
<p>Participants will learn a complete set of steps for conducting a differential binding analysis of data from a ChIP-seq or related DNA enrichment experiment within <em>Bioconductor</em>.</p>
<div class="section level2">
<h2 id="learning-goals">Learning goals<a class="anchor" aria-label="anchor" href="#learning-goals"></a>
</h2>
<ul>
<li>Gain an understanding of the processing flow for performing quantitative analysis of data from ChIP-seq and related DNA enrichment assays</li>
<li>Be able to filter and count aligned reads to prepare them for analysis</li>
<li>Appreciate the challenges inherent in normalizing these data</li>
<li>Complete a differential binding analysis</li>
</ul>
</div>
<div class="section level2">
<h2 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h2>
<ul>
<li>Load experimental data into <em>DiffBind</em>
</li>
<li>Filter aligned reads using blacklists, greylists, and mapping quality</li>
<li>Derive a consensus peak set and compute a matrix of read counts</li>
<li>Normalize read counts appropriately</li>
<li>Establish a design to model the data and one or more contrasts to test hypotheses based on the model</li>
<li>Report, plot, and export results for further analysis</li>
</ul>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/bioinformatics-core-shared-training/Quantitative-ChIPseq-Workshop/">Browse source code</a></li>
<li><a href="https://github.com/bioinformatics-core-shared-training/Quantitative-ChIPseq-Workshop/issues/new/choose">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing QuantitativeChIPseqWorkshop</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Rory Stark <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/https://orcid.org/0000-0002-1790-5469" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Rory Stark.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
