<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Quantitative_ChIPseq_Workshop • QuantitativeChIPseqWorkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Quantitative_ChIPseq_Workshop">
<meta property="og:description" content="This package encapsulates a workshop demonstrating the quantitative analysis (especially differential binding) of ChIP-seq, ATAC-seq and related assays using R packages (particularly DiffBind).">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">QuantitativeChIPseqWorkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Quantitative-ChIPseq-Workshop.html">Workshop: Quantitative analysis of ChiP-seq, ATAC-seq, and related DNA enrichment assays</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bioinformatics-core-shared-training/Quantitative-ChIPseq-Workshop">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<hr>
<p>title: “Quantitative-ChIPseq-Workshop” author: “Rory Stark” date: “November 25, 2020” output: html_document —</p>
<div id="quantitiative-analysis-of-chip-seq-atac-seq-and-related-dna-enrichment-assays" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#quantitiative-analysis-of-chip-seq-atac-seq-and-related-dna-enrichment-assays" class="anchor"></a>Quantitiative analysis of ChiP-seq, ATAC-seq, and related DNA enrichment assays.</h1></div>
</div>
<div id="instructor" class="section level1">
<h1 class="hasAnchor">
<a href="#instructor" class="anchor"></a>Instructor</h1>
<p>Rory Stark, Principal Scientist Computational Biology, University of Cambridge, Cancer Research UK Cambridge Institute</p>
<p><a href="mailto:rory.stark@cruk.cam.ac.uk" class="email">rory.stark@cruk.cam.ac.uk</a></p>
</div>
<div id="workshop-description" class="section level1">
<h1 class="hasAnchor">
<a href="#workshop-description" class="anchor"></a>Workshop Description</h1>
<p>This workshop will demonstrate the steps involved in performing a quantitative analysis of ChIP-seq data in <em>Bioconductor</em> (up to and including differential binding analysis), with some discussion of related assays such as ATAC-seq. Particular attention will be paid to processing of aligned reads, including blacklisting, greylisting, filtering for quality and duplication, and the particular challenges presented when normalizing these data. While the workshop follows the <em>DiffBind</em> package vignette, the use of a number of other <em>Bioconductor</em> packages is discussed, including <em>csaw</em>, <em>ChIPQC</em>, <em>edgeR</em>, <em>DESeq2</em>, and <em>GreyListChIP</em>.</p>
<div id="pre-requisites" class="section level2">
<h2 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites</h2>
<p>Attendees should be familiar with the following:</p>
<ul>
<li>Basic knowledge of R syntax</li>
<li>Familiarity with core <em>Bioconductor</em> classes, such as <em>GRanges</em>.</li>
<li>Familiarity with ChIP-seq or a related assay, including aligning sequencing reads.</li>
<li>Familiarity with generalized linear models (GLMs), especially as realized in the <em>edgeR</em> and/or <em>DESeq2</em> packages.</li>
</ul>
<p>Helpful background reading:</p>
<ul>
<li>Read through the <em>DiffBind</em> package vignette</li>
<li>Read through the <em>csaw</em> package vignette</li>
</ul>
<p>There is a book chapter which offers a detailed outline of the design and analysis of ChIP-seq experiments:</p>
<ul>
<li>Stark, Rory, and James Hadfield. “Characterization of DNA-protein interactions: design and analysis of ChIP-seq experiments.” <em>Field Guidelines for Genetic Experimental Designs in High-Throughput Sequencing.</em> Springer, Cham, 2016. 223-260.</li>
</ul>
<p>Email the instructor if you would like a copy.</p>
</div>
<div id="workshop-participation" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-participation" class="anchor"></a>Workshop participation</h2>
<p>A mini-lecture will be given at the beginning that will cover basic ideas behind ChIP-seq, comparision of commonly used tools for ChIP-seq data analysis. Then, a hands-on demo will be performed to demonstrate the downstream analysis with ChIPpeakAnno and trackViewer packages. Last will be a Q/A section.</p>
</div>
<div id="how-to-run-docker" class="section level2">
<h2 class="hasAnchor">
<a href="#how-to-run-docker" class="anchor"></a>How to run Docker</h2>
<ul>
<li>Install Docker (<a href="https://www.docker.com/" class="uri">https://www.docker.com/</a>)</li>
<li>Pull the Docker image from Docker hub: <code>docker pull crukcibioinformatics/quantitative_chip_workshop</code>
</li>
<li>Run the image: <code>docker run -e PASSWORD=yourpassword -p 8787:8787 crukcibioinformatics/quantitative_chip_workshop</code>
</li>
<li>Log in to RStudio at <a href="http://localhost:8787">http://localhost:8787</a> using username <code>rstudio</code> and password <code>yourpassword</code>. For Windows users, you also need to provide your IP address, you can find it using <code>docker-machine ip default</code>.</li>
<li>Inside the RStudio, run: <code><a href="https://rdrr.io/r/utils/browseVignettes.html">browseVignettes(package = "Quantitative-ChIPseq-Workshop")</a></code>
</li>
<li>You can click one of the links: “HTML”, “source”, “R code”</li>
</ul>
</div>
<div id="r--bioconductor-packages-used" class="section level2">
<h2 class="hasAnchor">
<a href="#r--bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h2>
<ul>
<li><em>DiffBind</em></li>
<li><em>csaw</em></li>
<li><em>ChIPQC</em></li>
<li><em>DESeq2</em></li>
<li><em>edgeR</em></li>
<li><em>GreyListChIP</em></li>
</ul>
</div>
<div id="time-outline" class="section level2">
<h2 class="hasAnchor">
<a href="#time-outline" class="anchor"></a>Time outline</h2>
<p>The workshop duration is 90 min. Approximate timing of activities:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Introduction</td>
<td>5m</td>
</tr>
<tr class="even">
<td>QC of DNA enrichment assays</td>
<td>5m</td>
</tr>
<tr class="odd">
<td>Peaks vs. Windows</td>
<td>5m</td>
</tr>
<tr class="even">
<td>
<em>DiffBind</em> Sample sheet</td>
<td>5m</td>
</tr>
<tr class="odd">
<td>Blacklists and Greylists</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Counting aligned reads</td>
<td>10m</td>
</tr>
<tr class="odd">
<td>Normalization I</td>
<td>5m</td>
</tr>
<tr class="even">
<td>GLM model design and contrasts</td>
<td>10m</td>
</tr>
<tr class="odd">
<td>Normalization II</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Reporting</td>
<td>5m</td>
</tr>
<tr class="odd">
<td>Plotting</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Exporting results for downstream analysis</td>
<td>5m</td>
</tr>
<tr class="odd">
<td>Conclusions</td>
<td>5m</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="workshop-goals-and-objectives" class="section level1">
<h1 class="hasAnchor">
<a href="#workshop-goals-and-objectives" class="anchor"></a>Workshop goals and objectives</h1>
<p>Participants will learn a complete set of steps for conducting a differential binding analysis of data from a ChIP-seq or related DNA enrichment experiment within <em>Bioconductor</em>.</p>
<div id="learning-goals" class="section level2">
<h2 class="hasAnchor">
<a href="#learning-goals" class="anchor"></a>Learning goals</h2>
<ul>
<li>Gain an understanding of the processing flow for performing quantitative analysis of data from ChIP-seq and related DNA enrichment assays</li>
<li>Be able to filter and count aligned reads to prepare them for analysis</li>
<li>Appreciate the challenges inherent in normalizing these data</li>
<li>Complete a differential binding analysis</li>
</ul>
</div>
<div id="learning-objectives" class="section level2">
<h2 class="hasAnchor">
<a href="#learning-objectives" class="anchor"></a>Learning objectives</h2>
<ul>
<li>Load experimental data into <em>DiffBind</em>
</li>
<li>Filter aligned reads using blacklists, greylists, and mapping quality</li>
<li>Derive a consensus peak set and compute a matrix of read counts</li>
<li>Normalize read counts appropriately</li>
<li>Establish a design to model the data and one or more contrasts to test hypotheses based on the model</li>
<li>Report, plot, and export results for further analysis</li>
</ul>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/bioinformatics-core-shared-training/Quantitative-ChIPseq-Workshop/">https://​github.com/​bioinformatics-core-shared-training/​Quantitative-ChIPseq-Workshop/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/bioinformatics-core-shared-training/Quantitative-ChIPseq-Workshop/issues/new/choose">https://​github.com/​bioinformatics-core-shared-training/​Quantitative-ChIPseq-Workshop/​issues/​new/​choose</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Rory Stark <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/https://orcid.org/0000-0002-1790-5469" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Rory Stark.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
